x/ecocredit/server/msg_server.go

```diff
@@ -91,13 +90,63 @@ func (s serverImpl) CreateBatch(goCtx context.Context, req *ecocredit.MsgCreateB
 		return nil, err
 	}
 
+	projectID := req.ProjectId
+	if req.ProjectId == "" {
+		projectSeqNo := s.projectInfoSeq.NextVal(ctx)
+		projectID = ecocredit.FormatProjectID(classInfo.ClassId, projectSeqNo)
+	}
```

Comment:
This is implemented as described in the proto file. We want to give users the ability to choose their own project id.

I was wondering the same thing in regards to locking the value but `NextVal` will persist the sequence. This might result in a single error that the id is not available but it would succeed on retry. Maybe we can try again until a valid id?

Reply:
> This might result in a single error that the id is not available but it would succeed on retry.

It won't, because the transaction will be reverted. So this can essentially lock the _create_project_ functionality for autogenerated IDs.

Reply:
One solution to this is to add a loop here (and charge gas) until we will find the next available sequence number. Moreover, checking if this is OK should be done with `Has`, not `Create` because of the overhead of `Create`.

Reply:
I think looping wouldn't be the most ideal solution here -- we could consider having a reserved character to use as a postfix in conjunction with the sequence number. 

For example using "-" or "~". When a project ID with collision is found, we can postfix with "-[sequence]". we would have to disallow this character in user defined project id creations.

Reply:
> For example using "-" or "~". When a project ID with collision is found, we can postfix like so:

It'll break project ID validation. see [c7a36f97](https://github.com/regen-network/regen-ledger/pull/647/files#diff-c7a36f97c7dfaeeb3cd82e21d461378b365674620add0ad9a97e2f3c60798c8fR137).

Reply:
> It'll break project ID validation. see c7a36f97.

we could just have 2 separate validation schemes, one for creation which excludes `-` and `~`, and then another for ID's that have already been created

Reply:
We need to increment the sequence anyway... so I don't think there is a value of escaping the sequence iterator.

Also, IMHO I would remove custom IDs at all and use only auto-generated one... but this is a business decision.

Reply:
I generally think the solution is to loop and generate an ID that doesn't collide and charge extra gas rather than an escape. There is a business use case for custom project IDs.